postproc-lib/lib/SP/Endurance/GraphGenerators.pm

```diff
@@ -3258,15 +3258,16 @@ sub generate_plot_display_state {
     );
 
     my @snapshot_states = map { $_->{'display_state'} } @$masterdb;
-    my @blanked = map {
-        defined $_->{'statefs'} ? $_->{'statefs'}->{'namespaces/Screen/Blanked'} : undef;
-    } @$masterdb;
+    my @blanked = map { int ( $_->{'blanked'} ) } @$masterdb;
 
     # Try to fill data gaps in snapshots that had no display events in journal.
     for (my $i = @snapshot_states - 1; $i > -1; --$i) {
         my $state = $snapshot_states[$i];
         if (!defined $state->{'on_percent'}) {
             # Journal wasn't available in the snapshot.
+            if (defined $blanked[$i]) {
```

Comment:
Is this post processing? Or recording in device? Is there history / possibility for having such?

Just thinking that mce keeps tabs on how much time has been spent in each display state -> exact values are available without journal parsing. If stats at previous sample time are kept it should be possible to calculate exact delta too.

Caveat: "undefined" == uptime until mce startup -> restarting mce = all history goes "undefined"

Example  (note that there is "machine" variant for easier parsing too)

    mcetool --get-display-stats=human
    UNDEF          00:00:09.084, 0 times
    OFF         2d 16:05:11.744, 2 times
    LPM_OFF        00:00:10.016, 2 times
    LPM_ON         00:00:13.319, 3 times
    DIM            00:00:00.000, 0 times
    ON          1d 07:30:01.609, 3 times
    POWER_UP       00:00:01.054, 2 times
    POWER_DOWN     00:00:00.805, 2 times

Reply:
So it looks like this could replace all of the current processing? The graph is currently rendered with one bar per snapshot showing the percentage of time the screen was on during that snapshot (the use of the "blanked" state is a poor fallback).

So, could we replace all of that with `on_percent = ON/(OFF + ON)`?

Reply:
An example graph for reference, generated from 5 snapshots using the existing method:
![2000_display_time_in_state](https://user-images.githubusercontent.com/1446122/77734645-bfa15e00-7011-11ea-9080-38425c709411.png)

Reply:
@rantala, since you're using this in the context of a machine without mce, we'd need to leave in the journal approach as a fallback, if I understand correctly?

Reply:
The display stats are not really interesting for server machines. Feel free to drop it.

Reply:
(But it can be useful to keep compatibility for older sp-endrance data, if you want to regenerate graphs)

Reply:
@llewelld What is it that we are looking for in journal for those display stats? If it is based on mce activity and thus maemo/meego/mer/sailfish specific -> we might skip gathering/processing display state data if mce and/or mcetool binaries are not present. That could avoid noise in e.g. server systems mentioned by @rantala

Reply:
@llewelld Again, depends on what the on_percentage is meant to mean i.e.

    on_percentage = ON / (ON + OFF)

is fine in one sense, but it could also be something like:

    on_time = ON + DIM + LPM_ON + POWER_UP
    off_time = OFF + LPM_OFF + POWER_DOWN
    on_percent = on_time / (on_time + off_time)

also, if it needs to be / sample -> values reported by mce can't be used as is - delta to values from previous sample would be needed

Reply:
@spiiroin, it appears to be searching for `unsleepDisplay` and `sleepDisplay` messages generated by lipstick and then using their timestamps. See here: https://github.com/mer-tools/sp-endurance/blob/6b17a244e2bf2769f9757b917c4bd052cdeedc21/postproc-lib/lib/SP/Endurance/Parser.pm#L1342
So, that would be sailfish-specific I would say.

Reply:
> @llewelld Again, depends on what the on_percentage is meant to mean

Yes, unfortunately I don't know the answer to this. @mkosola, could you advise?

Reply:
The new version tried to incorporate the various suggestions for using mce:
1. Collects all of the data using `mcetool --get-display-stats=machine`.
2. If undefined hasn't increased by more than 100, uses a delta from the previous value.
3, Combines the values as described [above](https://github.com/mer-tools/sp-endurance/pull/23#discussion_r399114806).

It's hard for me to test this, or know if it's giving useful results, but... well, the code doesn't generate errors.